### PR TITLE
Set a unique client.id property in AbstractKafkaConfigProperties

### PR DESCRIPTION
**Pull request by calohmn**
Corresponding discussion:
https://github.com/eclipse/hono/pull/2440#discussion_r565363789

### CODE REVIEW

**Comment by b-abel** [comment]
I'm a little worried about name conflicts with the client ID. If a producer connects to a Kafka cluster and presents a client ID that is already known in the cluster, the producer will be rejected. For consumers, the situation is likely to be analogous. The mechanism in the PR uses the UUID to ensure that there are no conflicts between the IDs of different verticles (and application instances). But the question is how can we ensure that every time a client is terminated and reopened, a new client ID is generated. In `AbstractKafkaBasedMessageSender` this is done every time `start()` is invoked, with this PR it only happens at instantiation. There you can solve the situation by passing in the config object and invoking `getProducerConfig()` in `start()` each time. It would be nicer to have an API that enforces that, but at the moment I have no idea how this could be acchieved.

**Comment by calohmn** [comment]
How about using `KafkaProducerConfigProperties` instead of `Map<String, String>` in the `AbstractKafkaBasedDownstreamSender` and `AbstractKafkaBasedMessageSender` constructors then?

Then there should also be a `KafkaProducerFactory#getOrCreateProducer` method with a `KafkaProducerConfigProperties` instead of Map parameter.

**Comment by b-abel** [comment]
> How about using `KafkaProducerConfigProperties` instead of `Map<String, String>` in the `AbstractKafkaBasedDownstreamSender` and `AbstractKafkaBasedMessageSender` constructors then?
> 
> Then there should also be a `KafkaProducerFactory#getOrCreateProducer` method with a `KafkaProducerConfigProperties` instead of Map parameter.

Yes. That is what I meant. I would _replace_ the map parameter in all of the classes you mentioned. 

This does not solve the general problem that it is easy to use the interface of the config properties classes wrong, as you cannot _enforce_ that for each new Kafka client instance new properties are fetched. So, this should be at least documented very clearly at the getters (`getProducerConfig()` etc.).

**Comment by calohmn** [comment]
@b-abel I've amended the commit with the added `getProducerConfig` javadoc and added a 2nd commit containing the changes towards using `KafkaProducerConfigProperties` instead of a Map.

**Comment by calohmn** [comment]
(rebased)

File: clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/CachingKafkaProducerFactory.java [hunk 1]
```diff
@@ -86,6 +86,29 @@ public CachingKafkaProducerFactory(
...
+     * a new instance is created using the given factory and put to the cache.
+     * <p>
+     * The given config is ignored when an existing producer is returned.
+     *
+     * @param producerName The name to identify the producer.
+     * @param config The Kafka configuration with which the producer is to be created.
+     * @return an existing or new producer.
+     */
+    @Override
+    public KafkaProducer<K, V> getOrCreateProducer(final String producerName, final KafkaProducerConfigProperties config) {
```
Comment (b-abel) [inline clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/CachingKafkaProducerFactory.java:102]: FMPOV the other variant with the map parameter should be entirely be removed.

File: clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/KafkaProducerFactory.java [hunk 2]
```diff
@@ -67,6 +67,25 @@
...
+     * <p>
+     * When the producer is no longer needed, it should be closed to release the resources (like connections). NB: Do
+     * not close the returned producer directly. Instead, call {@link #closeProducer(String)} to let the factory close
+     * the producer and update its internal state.
+     *
+     * @param producerName The name to identify the producer.
+     * @param config The Kafka configuration with which the producer is to be created.
+     * @return an existing or new producer.
+     */
+    KafkaProducer<K, V> getOrCreateProducer(String producerName, KafkaProducerConfigProperties config);
```
Comment (b-abel) [inline clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/KafkaProducerFactory.java:87]: FMPOV the other variant with the map parameter should be entirely be removed.

Comment (calohmn) [reply clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/KafkaProducerFactory.java:68]: done

**Comment by calohmn** [comment]
@sophokles73 @b-abel Further comments here?

**Comment by sophokles73** [comment]
not from my side

**Comment by b-abel** [comment]
> @sophokles73 @b-abel Further comments here?

no

**Pull request merged**